src/metabase/driver/mysql.clj

```diff
@@ -86,6 +101,46 @@
 (defn- date-format [format-str expr] (hsql/call :date_format expr (hx/literal format-str)))
 (defn- str-to-date [format-str expr] (hsql/call :str_to_date expr (hx/literal format-str)))
 
+(defn- timezone-id->offset-str
+  "Get an appropriate timezone offset string for a timezone with `timezone-id`. MySQL only accepts these offsets as
+  strings like `-8:00`.
+
+      (timezone-id->offset-str \"US/Pacific\") ; -> \"-08:00\"
+
+  Returns `nil` if `timezone-id` is itself `nil`."
+  [^String timezone-id]
+  (when timezone-id
+    (time/unparse (time/formatter "ZZ" (t/time-zone-for-id timezone-id)) (t/now))))
+
+;; MySQL doesn't seem to correctly want to handle timestamps no matter how nicely we ask. SAD! Thus we will just
+;; convert them to appropriate timestamp literals and include functions to convert timezones as needed
+(defmethod sqlqp/->honeysql [MySQLDriver Date]
+  [_ date]
+  (let [timestamp-literal          (hx/literal (u/format-date :date-hour-minute-second-ms date))
+        report-timezone-offset-str (timezone-id->offset-str (driver/report-timezone))
+        system-timezone-offset-str (timezone-id->offset-str (.getID (TimeZone/getDefault)))]
+    (if (and report-timezone-offset-str
+             (not= report-timezone-offset-str system-timezone-offset-str))
+      ;; if we have a report timezone we want to generate SQL like convert_tz('2004-01-01T12:00:00','-8:00','-2:00')
+      ;; to convert our timestamp from system timezone -> report timezone.
+      ;; See https://dev.mysql.com/doc/refman/5.7/en/date-and-time-functions.html#function_convert-tz
+      ;; (We're using raw offsets for the JVM timezone instead of the timezone ID because we can't be 100% sure that
+      ;; MySQL will accept either of our timezone IDs as valid.)
+      ;;
+      ;; Note there's a small chance that report timezone will never be set on the MySQL connection, if attempting to
+      ;; do so fails because the ID is valid; if the report timezone is different from the MySQL database's timezone,
+      ;; this will result in the `convert_tz()` call below being incorrect. Unfortunately we don't currently have a
+      ;; way to determine that setting a timezone has failed for the current query, since it actualy is attempted
+      ;; after the query is compiled. Hopefully situtations where that happens are rare; at any rate it's probably
+      ;; preferable to have timezones slightly wrong in these rare theoretical situations, instead of all the time, as
+      ;; was the previous behavior.
+      (hsql/call :convert_tz
+        timestamp-literal
+        (hx/literal system-timezone-offset-str)
+        (hx/literal report-timezone-offset-str))
+      ;; otherwise if we don't have a report timezone just we just want SQL like timestamp('2004-01-01T12:00:00')
+      (hsql/call :timestamp timestamp-literal))))
```

Comment:
not sure if it actually makes sense just to continue using `java.sql.Timestamp` objects here if we're not doing any offsets

Reply:
I think it should be ok to leave the timestamp as-is. If we don't have a report-timezone, we just attach the timestamp directly without a calendar associated with it. That leaves no report-timezone path the same as it was before.